filogic: add support for GL.iNet GL-MT6000

zhaojh329 opened the pull request:
Hardware specification:
* SoC: MediaTek MT7986A 4x A53
* Flash: 8GB EMMC
* RAM: 1GB DDR4
* Ethernet:
  * 2x2.5G RJ45 port (RTL8221B)
  * 4x1G RJ45 ports (MT7531AE)
* WLAN:
  * 2.4GHz: MT7976GN 4T4R
  * 5GHz: MT7976AN 4T4R
* Button: Reset
* LED: 1 x dual color LED
* USB: 1 x USB 3.0
* Power: DC 12V 4A
* UART: 3V3 115200 8N1 (Pinout: GND TX RX VCC)
* JTAG: 9 PIN

If you want to use u-boot from OpenWrt, you can upgrade it safely.
* bl2: openwrt-mediatek-filogic-glinet_gl-mt6000-preloader.bin
* fip: openwrt-mediatek-filogic-glinet_gl-mt6000-bl31-uboot.fip

`openwrt-mediatek-filogic-glinet_gl-mt6000-squashfs-factory.bin` is used in OpenWrt's u-boot.


![image](https://github.com/openwrt/openwrt/assets/28673295/431b9846-228d-455b-8fba-e123e690dc1f)

![image](https://github.com/openwrt/openwrt/assets/28673295/4750e15d-9169-44d1-bee2-4783daa1e7fc)

aiamadeus committed:
mt76: drop default eeprom file for mt7986-firmware

The mt76 driver usually reads the eeprom on the mtd partition at dts.
For emmc device we need to use caldata_extract script to read the
eeprom. However, the default eeprom file breaks the caldata script
execution, so remove it.

Signed-off-by: Chukun Pan <amadeus@jmu.edu.cn>

aiamadeus commented:
![image](https://github.com/openwrt/openwrt/assets/42570690/2203ab8a-c5c2-401a-93b2-9c72bbbfcc54)?

obsy commented on package/base-files/files/bin/config_generate:
```diff
@@ -162,7 +162,7 @@ generate_network() {
 			local ipad
 			case "$1" in
-				lan) ipad=${ipaddr:-"192.168.1.1"} ;;
+				lan) ipad=${ipaddr:-"192.168.8.1"} ;;
```
This should not be changed. 192.168.8.1 is gl-inet specific.

zhaojh329 commented:
> ![image](https://user-images.githubusercontent.com/42570690/265694417-2203ab8a-c5c2-401a-93b2-9c72bbbfcc54.png)?

awkward.

dangowrt commented on package/boot/uboot-mediatek/patches/436-add-glinet-mt6000.patch:
```diff
@@ -0,0 +1,250 @@
++
++/ {
++	#address-cells = <1>;
++	#size-cells = <1>;
```
Both lines above should not be needed in board dts file

dangowrt commented on package/boot/uboot-mediatek/patches/436-add-glinet-mt6000.patch:
```diff
@@ -0,0 +1,250 @@
++	#address-cells = <1>;
++	#size-cells = <1>;
++	model = "GL.iNet GL-MT6000";
++	compatible = "mediatek,mt7986", "mediatek,mt7986-emmc-rfb";
```
Should be `glinet,gl-mt6000` as first compatible string (most specific) with the others following in the order from most specific to least specific (ie. `mediatek,mt7986` goes last)

dangowrt commented on target/linux/generic/pending-5.15/731-net-phy-realtek-support-interrupt-of-RTL8221B.patch:
This patch should also go into `target/linux/generic/pending-6.1`.

dangowrt reviewed (changes requested):
Looks quite good already, only some small/cosmetic changes are needed.

dangowrt commented on target/linux/generic/pending-6.1/741-net-phy-realtek-support-interrupt-of-RTL8221B.patch:
```diff
@@ -0,0 +1,63 @@
++{
++	int irq_status;
++
++	irq_status = phy_read_mmd(phydev, RTL8221B_MMD_PHY_CTRL, 0xa4d4);
```
Couldn't this be replaced by a call to `rtl8221b_ack_interrupt` defined above?

zhaojh329 replied:
Yeah, you are right.

zhaojh329 commented:
I found a problem.
```
root@OpenWrt:~# [   32.487561] usb_vbus: disabling
```

zhaojh329 commented:
> I found a problem.
> 
> ```
> root@OpenWrt:~# [   32.487561] usb_vbus: disabling
> ```

The usb not working in master, but in 23.05 is OK.

mrkiko commented:
Hi!
May I ask you if you can add installation instructions / recovery method?

Thanks a lot!

Enrico

zhaojh329 commented:
> 

Installation instructions / recovery method for devices of GL.iNet is well-known.

mrkiko commented:
The "surrounding" changes looks good to me; I haven't checked the DTS tough.
Very apreciated you added u-boot support as well.

zhaojh329 commented:
> I'll add a bootmenu for the u-boot later.

Done.

zhaojh329 commented on package/boot/uboot-mediatek/patches/436-add-glinet-mt6000.patch:
```diff
@@ -0,0 +1,270 @@
++bootfile_fip=openwrt-mediatek-filogic-glinet_gl-mt6000-bl31-uboot.fip
++bootfile_kernel=glinet_gl-mt6000-kernel.bin
++bootmenu_confirm_return=askenv - Press ENTER to return to menu ; bootmenu 60
++bootmenu_title=      *** U-Boot Boot Menu ***
```
Good.

zhaojh329 commented:
@dangowrt 
I still need a menu to upgrade rootfs.
I need some way to convert size in bytes to block count to upgrade rootfs. Do you have any suggestions?

dangowrt commented:
The changes themselves look all good now, however, it would be nice to split this into several commits:
 * commit adding `*net-phy-realtek-support-interrupt-of-RTL8221B.patch`
 * commit add U-Boot build for MT6000
 * commit adding uboot-envtools settings for MT6000
 * commit add DTS and image build for MT6000

zhaojh329 commented:
> The changes themselves look all good now, however, it would be nice to split this into several commits:
> 
> * commit adding `*net-phy-realtek-support-interrupt-of-RTL8221B.patch`
> * commit add U-Boot build for MT6000
> * commit adding uboot-envtools settings for MT6000
> * commit add DTS and image build for MT6000

OK.

zhaojh329 commented:
@dangowrt 
The mt76 driver usually reads the eeprom on the mtd partition at dts. For emmc device we need to use caldata_extract script to read the eeprom. However, the default eeprom file breaks the caldata script execution.

https://github.com/openwrt/openwrt/blob/main/package/kernel/mt76/Makefile#L590

zhaojh329 committed:
generic: net: phy: realtek: add interrupt support for RTL8221B

This commit introduces interrupt support for RTL8221B.

Signed-off-by: Jianhui Zhao <zhaojh329@gmail.com>

dangowrt commented:
Yes, I know and agree. In the future we should have nvmem-cells attached to MMC hardware or software partitions, but for now adn Linux 5.15 in OpenWrt we need a work-around in user-space. So yeah, those default EEPROM files should each go into a package of its own...

zhaojh329 committed:
filogic: add support for GL.iNet GL-MT6000

Hardware specification:
* SoC: MediaTek MT7986A 4x A53
* Flash: 8GB EMMC
* RAM: 1GB DDR4
* Ethernet:
  * 2x2.5G RJ45 port (RTL8221B)
  * 4x1G RJ45 ports (MT7531AE)
* WLAN:
  * 2.4GHz: MT7976GN 4T4R
  * 5GHz: MT7976AN 4T4R
* Button: Reset
* LED: 1 x dual color LED
* USB: 1 x USB 3.0
* Power: DC 12V 4A
* UART: 3V3 115200 8N1 (Pinout: GND TX RX VCC)
* JTAG: 9 PIN

If you want to use u-boot from OpenWrt, you can upgrade it safely.
* bl2: openwrt-mediatek-filogic-glinet_gl-mt6000-preloader.bin
* fip: openwrt-mediatek-filogic-glinet_gl-mt6000-bl31-uboot.fip

`openwrt-mediatek-filogic-glinet_gl-mt6000-squashfs-factory.bin` is used in OpenWrt's u-boot.

Signed-off-by: Jianhui Zhao <zhaojh329@gmail.com>

aiamadeus commented:
Actually you should use pick or am my commit: https://github.com/aiamadeus/openwrt/commit/e8261f7cd59b5d67af9a76b4d45f6790569a9e7e

zhaojh329 commented:
> Actually you should use pick or am my commit: [aiamadeus@e8261f7](https://github.com/aiamadeus/openwrt/commit/e8261f7cd59b5d67af9a76b4d45f6790569a9e7e)

Done.

dangowrt commented:
picked to main branch https://github.com/openwrt/openwrt/commit/fe10f9743935d6986e80e7cb082469e6bc5a03f0

DRCRecoveryData commented:
> Hardware specification:
> 
> * SoC: MediaTek MT7986A 4x A53
> * Flash: 8GB EMMC
> * RAM: 1GB DDR4
> * Ethernet:
>   
>   * 2x2.5G RJ45 port (RTL8221B)
>   * 4x1G RJ45 ports (MT7531AE)
> * WLAN:
>   
>   * 2.4GHz: MT7976GN 4T4R
>   * 5GHz: MT7976AN 4T4R
> * Button: Reset
> * LED: 1 x dual color LED
> * USB: 1 x USB 3.0
> * Power: DC 12V 4A
> * UART: 3V3 115200 8N1 (Pinout: GND TX RX VCC)
> * JTAG: 9 PIN
> 
> If you want to use u-boot from OpenWrt, you can upgrade it safely.
> 
> * bl2: openwrt-mediatek-filogic-glinet_gl-mt6000-preloader.bin
> * fip: openwrt-mediatek-filogic-glinet_gl-mt6000-bl31-uboot.fip
> 
> `openwrt-mediatek-filogic-glinet_gl-mt6000-squashfs-factory.bin` is used in OpenWrt's u-boot.
> 
> ![image](https://user-images.githubusercontent.com/28673295/270529243-431b9846-228d-455b-8fba-e123e690dc1f.png)
> 
> ![image](https://user-images.githubusercontent.com/28673295/270529491-4750e15d-9169-44d1-bee2-4783daa1e7fc.png)

Can i get this dev board for testing the firmware ?
